### DISCUSSION TITLE
How would I use Persistence ReadJournal from the current running actor system

### DISCUSSION THREAD

**Discussion by andre-steenbergen** [akkadotnet/akka.net#7188] [Q&A] [answered]
When I am going through the documentation I see https://getakka.net/articles/persistence/persistence-query.html specific SQL based ReadJournals. I am trying to build an Persistence Actor without knowledge of the persistence layer used.

For testing for example I am using the In Memory Journal using "Persistence.Instance.Apply(Sys).JournalFor(string.Empty);". Is their any way without to get a ReadJournal from the currently enabled Persistence provider?

I am trying to build an actor that processes all events and based on logic build a command into another actor. Basically a rules engine to know what commands to provide actors doing basic logic.

Is there a way to get this working? Context.System.GetReadJournal (or something like that)?

**Comment by jaydeboer** [comment]
I am not sure this is exactly the same, but I had an actor once that would behave differently based on the type of persistence.  I was using: `Context.System.ReadJournalFor<ICurrentEventsByTagQuery>(settings.PersistenceReadJournalId);` to get the journal out by what was specified in config.  This worked well for me for switching between InMemory for tests and MongoDB for the live system.

**Reply by andre-steenbergen** [reply]
I'll try thx

**Comment by andre-steenbergen** [comment]
This is what I came up with at this moment. I need to debug if this works, but I would say it should ... ?


```
   protected override void ConfigureAkka(AkkaConfigurationBuilder builder, IServiceProvider provider)
    {
        //https://github.com/akkadotnet/akka.net/discussions/6789
        //discussion how to add the ReadJournal
        
        builder
            .WithInMemoryJournal()
            .WithInMemorySnapshotStore()
            .StartActors((system, registry) =>
            {
                
                var journalProvider = new Akka.Persistence.Query.InMemory.InMemoryReadJournalProvider(
                    system.AsInstanceOf<ExtendedActorSystem>(),
                    Persistence.DefaultConfig()
                    );

                var journal = journalProvider.GetReadJournal();
                CreateActer(system, registry, journal);
            });
    }
```

**Comment by andre-steenbergen** [comment, marked as answer]
This worked for me. The actor props is taking in a `IAllEventsQuery journal` which is provider agnostic

```
builder
            .WithInMemoryJournal()
            .WithInMemorySnapshotStore()
            .StartActors((system, registry) =>
            {

                var journalProvider = new InMemoryReadJournalProvider(
                    system.AsInstanceOf<ExtendedActorSystem>(),
                    InMemoryReadJournal.DefaultConfiguration().GetConfig("akka.persistence.query.journal.inmem")
                );

                var journal = journalProvider.GetReadJournal() as InMemoryReadJournal;
                
                var messageRouter = system.ActorOf(MessageRouter.GetProps());
                registry.Register<MessageRouter>(messageRouter);

                var approver = system.ActorOf(Approver.GetProps(journal));
                registry.Register<Approver>(approver);
            });
```
